src/client/opamPinCommand.ml

```diff
@@ -703,16 +703,26 @@ let unpin_one st nv =
     let st = OpamSwitchState.update_package_metadata nv opam st in
     { st with available_packages }
 
-let unpin st names =
+let unpin st names_opt_v =
   log "unpin %a"
-    (slog @@ OpamStd.List.concat_map " " OpamPackage.Name.to_string) names;
-  List.fold_left (fun st name ->
+    (slog @@ OpamStd.List.concat_map " " OpamPackage.to_string_opt_version) names_opt_v;
+  List.fold_left (fun st (name, version) ->
       OpamFilename.rmdir
         (OpamPath.Switch.pinned_package st.switch_global.root st.switch name);
       OpamFilename.rmdir
         (OpamPath.Switch.Overlay.package
            st.switch_global.root st.switch name);
-      match OpamPinned.package_opt st name with
+      let nv = match OpamPinned.package_opt st name, version with
+        | Some nv, Some v when OpamPackage.Version.equal (OpamPackage.version nv) v -> Some nv
+        | Some _, Some v ->
+          OpamConsole.error_and_exit `Bad_arguments "%s is pinned but not to version %s."
+            (OpamPackage.Name.to_string name) (OpamPackage.Version.to_string v)
+        | Some nv, None -> Some nv
+        | None, _ ->
+          OpamConsole.note "%s is not pinned." (OpamPackage.Name.to_string name);
+          None
```

Comment:
Is there a reason to not have this check in `OpamCommands.pin` directly ?

Reply:
I don't know, it was there before. I didn't change this part of the code

Reply:
I move the note back to the original position to make it clearer.

Reply:
Maybe i selected too much lines :) I wasn't referring to that note, but to the version check (and its consequence to update the api).

Reply:
ah i see. Mmh, I'll have a look to see if it's easier this way.

Reply:
Plus it will permit to filter packages (names) to unpin, removing ones with wrong version instead of failing if only have a wrong version.

Reply:
Yep. I don't know why i did it this way in the first place but your suggestion is indeed much simpler and better. This is now done